aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -132,11 +137,13 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
 } // anonymous namespace
 
 Tensor& mm_out_cuda(Tensor& result, const Tensor& self, const Tensor& mat2) {
+  globalContext().alertCuBLASConfigNotDeterministic();
   result.resize_({ self.size(0), mat2.size(1) });
   return addmm_out_cuda_impl(result, result, self, mat2, 0, 1);
 }
 
 Tensor mm_cuda(const Tensor& self, const Tensor& mat2) {
+  globalContext().alertCuBLASConfigNotDeterministic();
```

Comment:
Initially, I only added alerts in the internal CuBLAS wrapper functions, CUDABlas.cpp and THCBlas.cu. But then when I created and ran tests for a handful of the torch operations that use these functions (like `torch.mm`, `torch.dot`, etc), I was getting some CUDA memory access errors when I ran the tests back to back.

The problem seemed to be that the error was being thrown halfway through some operations and memory could sometimes be left in an unsafe state. So I had to call the alert function here instead, before the operations have a chance to do anything to the memory.

I think we should keep the alerts in the CuBLAS wrappers though, so that we automatically have error coverage over every operation that calls them.

But I wonder if I should continue to add alerts and tests for each existing torch operation that calls the CuBLAS wrappers. I feel like that might be overkill, but not sure.

Reply:
Oof, that's not very nice. In principle it should be safe to unwind the stack at any given point if we are using proper C++ destructors, I wonder if something is still using legacy behavior. It's possible this is related specifically to the `th_` wrappers. This is probably expedient for now.

Reply:
I'll pin down exactly what the issue was.

Reply:
It was simpler than I assumed. `at::native::dot_cuda()` changes the CuBLAS pointer mode with `cublasSetPointerMode()` before calling `at::cuda::blas::dot()`. It restores the previous setting after the call, but only if the call doesn't throw an error. So I had to just put the `at::cuda::blas::dot()` call in a try-catch. If there's an error, it will now restore the pointer mode and rethrow.

Reply:
It's even simpler than that, dot is always deterministic, regardless of the workplace setting.

Reply:
Oh I didn't realize that. How do you know that @ngimel? Are any of the other CuBLAS functions always deterministic regardless of the setting?

Reply:
I'll talk to nvidia to get the exact list, I strongly suspect that it's only gemms/batched gemms.

Reply:
@ptrblck says that only gemv and gemm are affected (also syrk/herk/trsm but we are not using them in pytorch. Magma might potentially be heavy on those, and it's pretty hard to know which magma routines are affected and which are not. On the bright side, for "good enough" according to @t-vi, we don't have to care about magma).

Reply:
"gemv and gemm" meaning only the non-batched versions? Or are the batched ones affected too?

Reply:
Only non-batched versions, batched are not affected. What if batch is 1, you might ask? Well, I don't know, I'm just relaying the message. @ptrblck ?

Reply:
Batched versions of gemv and gemm with a batch size of 1 can be non-deterministic, so we would need to disallow them.

Reply:
Great, thanks for letting me know! So I'll remove the alerts for everything but gemv and gemm, batched and unbatched. I think there are a couple questions to answer about how to handle the batch size 1 case though. It seems like these are our options:

1. Alert batched gemv and gemm always
2. Alert batched gemv and gemm only if batch size is 1
    a. Add a message to the alert explaining that it's nondeterministic because the batch size is 1
    b. Don't add the message

I think 2.a. might be the best, but I'm not sure.

Reply:
I removed the errors for ger and dot, so now only mm, mv, and batched multiplies throw errors. This is option 1 from my above comment. Please let me know if it would be better to actually check the batch size and provide more detail in the error message.